PhotoReader/kotlin/src/main/java/com/huawei/mlkit/sample/photoreader/util/TasksExtension.kt

```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.huawei.mlkit.sample.photoreader.util
+
+import com.huawei.hmf.tasks.Task
+import kotlinx.coroutines.suspendCancellableCoroutine
+import java.util.concurrent.CancellationException
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+
+/**
+ * wrap task onCompletion into a suspendable function so it can be used in a coroutine context
+ */
+suspend fun <T> Task<T>.await(): T {
+
+    // fast path
+    if (isComplete) {
+        val e = exception
+        return if (e == null) {
+            if (isCanceled) {
+                throw CancellationException("Task $this was cancelled normally.")
+            } else {
+                @Suppress("UNCHECKED_CAST")
+                result as T
+            }
+        } else {
+            throw e
+        }
+    }
+
+    return suspendCancellableCoroutine { cont ->
```

Comment:
too bad here we cannot cancel the underlying Task in case the coroutine is canceled - a not so elegant way would be to check the task is an instance of CancellableTask and cancel then with `cont.invokeOnCancellation { cancelableTask.cancel() }`

Reply:
Yep you are right, I actually missed that one. Still the underlying library must support cancelation to make it work anyway, since Task doesn't offer any cancellation function, add `invokeOnCancellation` wont bring any value. :(

Reply:
Maybe check if the task is of type `CancellableTask` and then call cancel, but instance checks are not the nicest.

Reply:
just checked, there is not a CancellableTask os something similar in classpath :D